### ISSUE TITLE
Squiz.ControlStructures.ForLoopDeclaration confused on empty expressions with spacing = 1

### ISSUE THREAD

**Issue by umherirrender** [squizlabs/PHP_CodeSniffer#2366]
The php documentation - http://php.net/manual/en/control-structures.for.php - says that the for loop can have all its expressions empty
Using `Squiz.ControlStructures.ForLoopDeclaration` with `requiredSpacesAfterOpen = 1` and `requiredSpacesBeforeClose = 1` confused the sniff.

```
for ( ; ; ) {
}
```
The code example triggers `SpacingBeforeFirst` and `SpacingBeforeSecond`

I am not sure, if `for ( ;; )` or `for ( ; ; )` would be the prefered way when using the spacing version.

The problem in my codebase is that only the first expression is empy and `SpacingBeforeFirst` is triggered. But it would be nice to get both cases fixed.

Thanks.

**Comment by jrfnl** [comment]
@umherirrender Could you please test open PR #2286 ? If I remember correctly, I've already addressed your concerns in that PR.

**Comment by umherirrender** [comment]
> 
> 
> @umherirrender Could you please test open PR #2286 ? If I remember correctly, I've already addressed your concerns in that PR.

You patch set looks good and when the first expression is empty, now it does not report anything. That is good.

### report errors on the line they occur
works good

### improve the accuracy of the SpaceAfterOpen/SpaceBeforeClose error messages
The error message now include "newline". Having a setting "ignoreNewlines" to ignore such messages would be nice and also inline with some other sniffs (maybe in a follow up)

### improve the semicolon spacing determination
Same as before, works but a new setting would be nice (maybe in a follow up)

### others
In case of
```
for ( ; ; ) {
}
```
it still reports the space before the second semicolon. In my codebase this way is more prefered than no spacing between the empty semicolon, maybe it needs a setting to set the spacing between two empty semicolons here

### Conclusion
Good begin of improvements to the sniff
Just for note, that the first and last expression can have commas, which are not adjust for whitespaces yet. But that is also something for a new patch set.
The improvments for the newline detects also needed in ForEachLoopDeclarationSniff.php

Good work!
